h/traversal/group.py

```diff
@@ -1,86 +1,55 @@
+    # Any logged in user may create a group
+    __acl__ = [(Allow, role.User, "create")]
 
     def __init__(self, request):
         super().__init__(request)
         self.group_service = request.find_service(name="group")
 
     def __getitem__(self, pubid_or_groupid):
-        group = self.group_service.fetch(pubid_or_groupid)
-        if group is None:
-            raise KeyError()
-        return group
+        # Group could be `None` here!
+        return GroupContext(group=self.group_service.fetch(pubid_or_groupid))
 
 
-class GroupUpsertRoot(RootFactory):
+class GroupRequiredRoot(GroupRoot):
     """
-    Root factory for group "UPSERT" API
+    Root factory for routes dealing with groups which must exist.
 
-    This Root can support a route in which the traversal's ``__getitem__``
-    will attempt a lookup but will not raise if that fails.
-
-    This is to allow a single route that can accept and update an existing group
-    OR create a new one.
+    FIXME: This class should return GroupContext objects, not Group objects.
     """
 
-    __acl__ = GroupRoot.__acl__
-
-    def __init__(self, request):
-        super().__init__(request)
-        self._group_root = GroupRoot(request)
-
     def __getitem__(self, pubid_or_groupid):
-        try:
-            group = self._group_root[pubid_or_groupid]
-        except KeyError:
-            group = None
+        group_context = super().__getitem__(pubid_or_groupid)
+        if group_context.group is None:
+            raise KeyError()
 
-        return GroupUpsertContext(group=group, request=self.request)
+        return group_context.group
 
 
-class GroupUpsertContext:
-    """Context for group UPSERT"""
+@dataclass
+class GroupContext:
+    """Context for a single (optional) group."""
 
-    def __init__(self, group, request):
-        self._request = request
-        self.group = group
+    group: Optional[Group] = None
 
     def __acl__(self):
-        """
-        Get the ACL from the group model or set "upsert" for all users in absence of model
-
-        If there is a group model, get the ACL from there. Otherwise, return an
-        ACL that sets the "upsert" permission for authenticated requests that have
-        a real user associated with them via :attr:`h.auth.role.User`.
-
-        The "upsert" permission is an unusual hybrid. It has a different meaning
-        depending on the upsert situation.
-
-        If there is no group associated with the context, the "upsert" permission
-        should be given to all real users such that they may use the UPSERT endpoint
-        to create a new group. However, if there is a group associated with the
-        context, the "upsert" permission is managed by the model. The model only
-        applies "upsert" for the group's creator. This will allow the endpoint to
-        support updating a specific group (model), but only if the request's
-        user should be able to update the group.
```

Comment:
I don't think it's beneficial to explain the contents of the ACLs in another file partially because it's long, and partially because it's easy for this to end up being a lie. The rest I think has been summarised below in fewer words.